librubyfmt/src/render_queue_writer.rs

```diff
@@ -18,21 +18,59 @@ impl RenderQueueWriter {
 
     pub fn write<W: Write>(self, writer: &mut W) -> io::Result<()> {
         let mut accum = Intermediary::new();
+        let mut heredocs = vec![];
         #[cfg(debug_assertions)]
         {
             debug!("first tokens {:?}", self.tokens);
         }
-        Self::render_as(&mut accum, self.tokens);
+        Self::render_as(&mut accum, self.tokens, &mut heredocs);
         Self::write_final_tokens(writer, accum.into_tokens())
     }
 
-    fn render_as(accum: &mut Intermediary, tokens: Vec<ConcreteLineTokenAndTargets>) {
+    fn render_as(accum: &mut Intermediary, tokens: Vec<ConcreteLineTokenAndTargets>, heredocs: &mut Vec<HeredocString>) {
         for next_token in tokens.into_iter() {
             match next_token {
                 ConcreteLineTokenAndTargets::BreakableEntry(be) => {
-                    Self::format_breakable_entry(accum, be)
+                    Self::format_breakable_entry(accum, be, heredocs)
+                }
+                ConcreteLineTokenAndTargets::HeredocString(hs) => {
+                    heredocs.push(hs);
+                    #[cfg(debug_assertions)]
+                    {
+                        debug!("render_as: got heredoc {}", heredocs.len())
+                    }
+                }
```

Comment:
The match branches for `HeredocString` here and `ConcreteLineToken` below are doing the majority of the work in this PR. At this point in the app, there is enough information available to know when to emit the heredocs, ie, it has been decided if the expression will be single or multiline.